napari/_qt/_tests/test_get_reader_choice.py

```diff
@@ -0,0 +1,88 @@
+from typing import Dict, Optional, Tuple
+
+from qtpy.QtWidgets import QWidget
+
+from napari._qt.qt_viewer import _get_reader_choice_for_file
+
+
+class MockQtReaderDialog:
+    """Dialog for user to select a reader plugin for a given file extension or folder"""
+
+    def __init__(
+        self,
+        pth: str = '',
+        parent: QWidget = None,
+        readers: "Dict[str, str]" = {},
```

Comment:
this can be unquoted.  The quotes are really there for [forward references](https://www.python.org/dev/peps/pep-0484/#id28).  I don't think mypy etc really care right now, but sometimes you don't get the right lint. 

```suggestion
        readers: Dict[str, str] = {},
```

I only flagged this here though there are other spots where you could unquote the types.  I'm approving anyway though so it's up to you.

Reply:
I've unquoted them now. I don't know exactly what prompted me to quote them, I think I've had instances of tests failing when they're not quoted... maybe because they're imported only `if TYPE_CHECKING`? Does that mean I should also remove the `if TYPE_CHECKING` imports @nclack? And, in what situation do we import types under the `if TYPE_CHECKING`

Reply:
[This](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING) is a pretty decent explanation.

If you're importing an expensive-to-import module (maybe like Qt) just to get its types, you can avoid the cost at runtime by using `TYPE_CHECKING`.  It's `False` normally, and only true when a tool like `mypy` is actually doing the type checking.

Reply:
they key thing here is whether or not you use `from __future__ import annotations` or not

so if you hide an import behind `if TYPE_CHECKING`, then you must either quote it, or use `from __future__ import annotations` (which forces the python parser to interpret _all_ type annotations as strings, according to [PEP 563](https://www.python.org/dev/peps/pep-0563/)) 

```python
# OK
if TYPE_CHECKING:
    import some_module

def f(x: 'some_module.SomeClass'): ...
```
```python
# OK
from __future__ import annotations
if TYPE_CHECKING:
    import some_module

def f(x: some_module.SomeClass): ...
```
```python
# NOT OK
# since when python tried to evaluate some_module.SomeClass at runtime
# it will not be in the global namespace
if TYPE_CHECKING:
    import some_module

def f(x: some_module.SomeClass): ...
```
```python
# OK, but superfluous, since some_module is available in the namespace
import some_module

def f(x: 'some_module.SomeClass'): ...
```

Reply:
Ok now I get it. So right now, I'm just importing the typing stuff directly, because even though one or two come from `qtpy` and may be expensive, there are other similarly expensive imports that are needed regardless of type checking. My feeling therefore is that there is no need to guard against them being expensive for no reason.

The only exception to that would be the `MockQtReaderDialog` [where we import `QWidget`](https://github.com/napari/napari/pull/3799/files/bf5d3a8a1cc6bd73536bbd7bb31ee0bc5bac8618#diff-82756d16b983ecc2be7257cfe44e270b830befd8b1bb26ea6218ba684f314b8fR8) even though it's only there as a mock parameter. So I'm going to put that behind type checking and quote the import.

Does this seem like a reasonable assessment @tlambert03?

Reply:
>  there are other similarly expensive imports that are needed regardless of type checking.

yep, you've got it right.  QWidget would be the only one.

It's all about namespaces.  There's effectively the "runtime" namespace and the "type checking" namespace. If you have to actually use the thing in the module during runtime, there's no need to put them behind TYPE_CHECKING (in fact, you can't, you'd have to also put them at the top). 

I generally do this:
1. if I'm _only_ importing a type for the sake of typing, i put it behind `TYPE_CHECKING`.  This is not only to avoid anything "expensive" ... since in most cases, it will have been imported into `sys.modules` anyway at runtime.  It's also for two very useful things:
   - avoid circular dependencies that might arise later.  We ran into this a lot with all of the async chunkloader stuff.  It tried to be thorough with type checking, which is great, but later became a real pain because every module imported every other module.  it was _very_ delicate to rearrange.
   - putting things in TYPE_CHECKING serves as nice documentation. I see that and say "ok, looks like this particular module isn't used here, but something in here is expecting one of these types"
2. If I've used TYPE_CHECKING and I only need the thing once or twice, then I'll just quote it (like `parent: 'QWidget'`)
3. If I'm quoting a _lot_ of things, then I tend to use `from __future__ import annotations`, which generally cleans things up.

There's two more things to be aware of.

You often need quotes or `__future__.annotations` for self-referencing classes:

```py
class T:
    @classmethod
    def create(cls) -> T:  # T is not defined... it must be quoted or use __future__.annotations
        return cls()
```

lastly, it's good to be aware that all these quoted "ForwardRef" type annotations _are_ trickier to use at runtime for, say, for pydantic or magicgui.  So, if you're certain that you _are_ going to want to analyze the type annotations at runtime, then you _might_ consider leaving them unquoted.

Reply:
Thank you @tlambert03 for the thorough explanation and additional info, as always!